README.md

# Compile ReCAP statistics

### why?

Ease of collaboration, of course

### say more

So, of course, MARC/SCSB XML is the most sensible lossless format
for the bibliographic metadata for the materials in ReCAP, but it
is _harrowing_ to work with, especially if unfamiliar with SAX,
XPath, regular expressions, etc...
Additionally, it's hierarchical nature is at odds with the "flat
spreadsheet" thing that most of us are used to.
Instead of all of us processing it in idiosyncratic ways, let's all
use one projection of these Marc XML data to flat file. We can all
collectively decide on the fields to include and how to include them.

It's also nice to have a nice version-controlled place to put this
code. In addition, we can (and will) set up a devops toolchain
to make this whole thing reproducible

### what, though?

So this is a multistep procedure that yields three spreadsheets
(tab-delimited, "NA" is null, etc...) meant for direct consumption:
  - `RECAP.dat.gz`\
    The projection of all the Marc XML data. As of time of writing,
    this includes 13.2 million items. (gzipped)
  - `transactions.dat`\
    Takes all the transaction data exported from LAS and, because there
    are overlaps in the time, removes the duplicates.
    At time of writing, this is 756,199 transactions from
    2017-06-20 to 2021-04-09.
  - `las-transaction-bib-info.dat`\
    In order to answer questions about what kind of materials we're
    borrowing from each other, the bib metadata table has to be
    joined with the transaction table. This is that.
    A few caveats... 20% of the barcodes in the transaction data do
    not have matches in the ReCAP XML data. _Why_ is a mystery.
    These records are still included, but there is no associated
    bib metadata, just the request date, owning and requesting
    institution, etc...
    Harvard, ILL, etc... transactions are ALSO included here. The
    HUL don't have any bib metadata, though, as they are not in the
    Marc XML data exports from SCSB, at time of writing

For space and privacy the input and output files are "git-ignored" and
not in this repository.

The (output) files are available on our
"ReCAP Assessment" shared drive under the directory "compiled-recap-stats"
Let the maintainer know if you need access or help with the files

The file names will have a date (corresponding to the date of the
MarcXML export/import, and may have a revision number.

---

### fields of `RECAP.dat`

  - __inst_has_item__\
    this is just the institution that owns the item

  - __barcode__\
    this is from 876\$p

  - __vol__\
    This is from 876\$3. This is really useful. It's usually NA for
    bibs that just have one item.

  - __numitems__\
    This is really useful, too. It's the number of items under each
    bib record

  - __scsbid__\
    From 001

  - __sharedp__\
    Whether it's "shared" or "open"

- __lang_code__\
    This is from characters 35 to 38 from the 008 (I think this is
    the most reliable)

  - __language__\
    A human-readable translation of the language code

  - __pubdate__\
    Characters 7 to 11 of the 008 field

  - __biblevel__\
    A human-readable translation of the seventh character of the
    MARC leader

  - __recordtype__\
    A human-readable translation of the sixth character of the
    MARC leader

  - __oclc__\
    More complicated than just the 035\$a... It only really allows
    035s that have (OCoCL) in them, but its more complicated than that

  - __lccn__\
    Taken from the 010\$a but then normalized/canonicalized to look pretty

  - __isbn__\
    Took all 020\$as, removed non-isbn-looking-ones, normalized and converted
    to ISBN13s, removed (now) duplicated numbers, and put them all in one
    cell separated by a semicolon

  - __original_isbn__\
    The original ISBN before normalization

  - __issn__\
    Same with ISBN but didn't convert to ISBN13 (obv)

  - __original_issn__\
    The original ISSN before normalization

  - __original_lccall__\
    The original contents of the LC call number field.
    Before any black magic

  - __lccall__\
    Probably the most complicated of the bunch...
    So we take 050\$a _and_ 090\$a. We preferentially chose the former
    but if it is missing, we take the later.
    _THEN_ a lot of call numbers appear to by missing but, upon closer
    inspection, they are often hiding in the local call number at
    field 852\$h (for institutions that are not NYPL). Those are looked
    through and, if the 050/090 thing is missing, the institution is _not_
    the NYPL, and there is a local call number that fits a regex pattern
    that strongly suggests that it is a LC call, it's brought into this
    column\
    _This is not perfect._ For example, Avery has books that appear to be
    LC call but actually aren't. There's nothing we can do about this
    until we join with, in this example, Columbia's complete ILS data.\
    For posterity, the LC call _without_ the procedure of taking it from
    the local call number is in a separate column called `original_lccall`

  - __localcallnum__\
    852\$h. Often Billings for NYPL and LC for CUL/PUL. Not always, though.
    See `lccall` directly above

  - __oh09__\
    The 090. For the NYPL it's our ILS' bibids. Idk what it is for CUL/PUL

  - __pubplace_code__\
    The country code found at character 15 to 18 of the 008

  - __pubplace__\
    A human-readable translation of the country code

  - __pubsubplace__\
    A regex-processed version of the 260\$a (usually publication city).
    If multiple, they are separated by semicolons

  - __leader__\
    MARC leader in full

  - __oh08__\
    008 in full

  - __dateoflastxaction__\
    They allege that the date of the last transaction can be found at
    0-8 of the 005 control field. I'm suspicious.

  - __title__\
    245\$a

  - __author__\
    100\$a

  - __topicalterms__\
    All the Library of Congress subject headers (650\$a for ind2 of 0)
    separated by a semicolon

  - __subject_classification__\
    Uses libbib to get LC subject classification from first letter
    of LC Call Number

  - __subject_subclassification__\
    Uses libbib to get LC subject _sub_ classification from all prefix
    letters in LC Call Number



CHANGELOG
2021-03-revision-2
- More hands-off dealings with dates of export
- Uses libbib's `fread/fwrite_plus_date`
- Step 2 uses assertr to check number of rows
- Step 3 no longer has to use lubridate
- Step 3 reads all las data automatically
- Step 4 uses assertr to check number of rows

2021-03-revision-1
- Uses libbib 1.3.1 to do more of the work
- Should fix problems with some LC Call -> Subject translations
- Better discriminating between local call numbers and lc calls
- utils in dependencies now part of libbib
- now writes out gzipped data files instead (for big files and intermediate files)
- writes targets in one place
- adds human-readable language name and place of publication name

2021-03
- adds LC Subject Classification and subclassification directly in data target
- uses libbib (on CRAN) instead of bibcodes.R
- fixed issue with export of LCCNs
- fixed issue with tabs and newlines in CUL MARCXml
- uses LAS transactions up to 2021-04-09
- remove garbage "tmp" variable
- various code/other/output improvements
- updated LICENSE and README

New items since last update (2020-05)
NYPL: 89,612
CUL:  40,774
PUL: 164,642  (issue in 2020-05 regarding open/shared items)

New number of transactions
205,849



step-1--parse-marc-xml/parse-all-marcxmls.lisp
#!/usr/local/bin/lispscript

; latest clix commit: 7c27c6d98e55

(use-xml-namespace  "http://www.loc.gov/MARC21/slim")

(load "xwalks.lisp")


(defmacro with-get-all (someval listofunctions &body body)
  (let ((dalist (mapcar (lambda (x) `(,x (,x ,someval))) listofunctions)))
    `(let ,dalist
       (let ((everything! (list ,@listofunctions)))
         ,@body))))


(defmacro defmarcxmlfield (thename theplain thegame &rest therest)
  (with-gensyms (tmp cc tmp2)
    `(let ((,cc (xpath-compile ,theplain)))
       (defun ,thename (adoc)
         (let ((it!
                 « (xpath adoc ,cc ,@therest :all nil :text t :compiled-p t)
                      or do (return-from ,thename nil) »))
           #? ,thegame)))))

; --------------------------------------------------------------- ;

(defmarcxmlfield leader             "leader" it!)
(defmarcxmlfield oh08       		    "controlfield[@tag='008']" it!)
(defmarcxmlfield oh09       		    "controlfield[@tag='009']" it!)
(defmarcxmlfield barcode    		    "datafield[@tag='876']/subfield[@code='p']" it!)
(defmarcxmlfield title      		    "datafield[@tag='245']/subfield[@code='a']" it!)
(defmarcxmlfield scsbid     		    "controlfield[@tag='001']" it!)
(defmarcxmlfield author     		    "datafield[@tag='100']/subfield[@code='a']" it!)
(defmarcxmlfield lccall1    		    "datafield[@tag='050']/subfield[@code='a']" it!)
(defmarcxmlfield lccall2    		    "datafield[@tag='090']/subfield[@code='a']" it!)
(defmarcxmlfield sharedp            "datafield[@tag='876']/subfield[@code='x']" it!)
(defmarcxmlfield lccn               "datafield[@tag='010']/subfield[@code='a']" it!)

(defmarcxmlfield localcallnum
  "datafield[@tag='852']/subfield[@code='h']"
  (~ra (str-trim it!) •\n• " "))

(defmarcxmlfield dateoflastxaction
  "controlfield[@tag='005']"
  (let ((yr (subseq it! 0 4))
        (mn (subseq it! 4 6))
        (dy (subseq it! 6 8)))
    (delim (list yr mn dy) :sep #\-)))

(defmarcxmlfield isbn
  "datafield[@tag='020']/subfield[@code='a']"
  (-<> it!
       (mapcar (lambda (x) (~ra x •^(\d+[Xx]?).*$• •\1•)) <>)
       (remove-duplicates <> :test #'equalp)
       (delim <> :sep #\;))
  :text t :all t)

(defmarcxmlfield issn
  "datafield[@tag='022']/subfield[@code='a']"
  (-<> it!
       (mapcar (lambda (x) (~ra x •^(\d+-?\d+[Xx]?).*$• •\1•)) <>)
       (remove-duplicates <> :test #'equalp)
       (delim <> :sep #\;))
  :text t :all t)

(defmarcxmlfield oclc
  "datafield[@tag='035']/subfield[@code='a']"
  (let ((res (remove-if-not (lambda (x) (~m x •^.OCoLC.•)) it!)))
    (when res (-<> res
                   (mapcar (lambda (x) (~r x •^\D+• "")) <>)
                   (remove-if-not (lambda (x) (~m x •^\d+$•)) <>)
                   (mapcar #'parse-integer <>)
                   (remove-duplicates <> :test #'eql)
                   (delim <> :sep #\;))))
  :text t :all t)

(defun lccall (something) (aif (lccall1 something) it! (lccall2 something)))

(defmarcxmlfield pubsubplace
  "datafield[@tag='260']/subfield[@code='a']"
  (-<> it!
       (mapcar (lambda (x) (~ra x •\s*[,:\.\]]+\s*$• "")) <>)
       (mapcar (lambda (x) (~ra x •^\[• "")) <>)
       (delim <> :sep #\;))
  :text t :all t)

(defmarcxmlfield topicalterms
  "datafield[@tag='650' and @ind2='0']/subfield[@code='a']"
  (delim (remove-duplicates it! :test #'equal) :sep #\;)
  :all t :text t)

;;; repeated :(
(defmarcxmlfield language
  "controlfield[@tag='008']"
  (subseq it! 35 38))

(defmarcxmlfield pubdate
  "controlfield[@tag='008']"
  (parse-integer (subseq it! 7 11)))

(defmarcxmlfield recordtype
  "leader"
  (-<> (subseq it! 6 7)
       (string-upcase <>)
       (find-symbol <>)
       {+record-type-xwalk+ <>}))

(defmarcxmlfield biblevel
  "leader"
  (-<> (subseq it! 7 8)
       (string-upcase <>)
       (find-symbol <>)
       {+bib-level-xwalk+ <>}))

(defmarcxmlfield pubplace
  "controlfield[@tag='008']"
  (subseq it! 15 18))


; --------------------------------------------------------------- ;


« (defvar /thedir/ Ø (cadr (cmdargs)))
    or die "no directory of xml files given" »

(defvar /allfiles/ (zsh (fn "find ~A -maxdepth 1 -type f | ack xml" /thedir/) :split t))
(defvar /intotal/ (length /allfiles/))


(defparameter /doc/ nil)


(defun get-item-info (node)
  (let ((barcodes (xpath node "datafield[@tag='876']/subfield[@code='p']" :all t :text t))
        (vol      (xpath node "datafield[@tag='876']/subfield[@code='3']" :all t :text t)))
    (setq vol (mapcar (lambda (x) (~ra (str-trim x) •\t• " " )) vol))
    (if (not (= (length barcodes) (length vol)))
      (loop for i in barcodes collect (list i NIL))
      (loop for i in barcodes
            for j in vol
            collect (list i (if (string= j "") NIL j))))))



(ft "~A~%" (delim (list "barcode" "vol" "numitems" "scsbid" "sharedp"
                        "lang_code" "pubdate" "biblevel" "recordtype"
                        "oclc" "lccn" "isbn" "issn" "lccall" "localcallnum"
                        "oh09" "pubplace_code" "pubsubplace" "leader" "oh08"
                        "dateoflastxaction" "title" "author"
                        "topicalterms")))



(for-each/list /allfiles/
  (progress index! /intotal/)
  (info "parsing file ~A~%" value!)
  (with-time
    (setq /doc/ (parse-xml-file value!))
    (info "parsing took ~A~%" (time-for-humans time!)))
  (with-time
    (for-each/list (xpath /doc/ "/collection/record")
      (let ((item-info (get-item-info value!)))
        (with-get-all value!
          (scsbid sharedp language  pubdate biblevel recordtype oclc lccn
           isbn issn lccall localcallnum oh09 pubplace pubsubplace leader
           oh08 dateoflastxaction title author topicalterms)
          (for-each/list item-info
            (ft "~A~C~A~C~A~C" (car value!) #\Tab (cadr value!) #\Tab
                               (length item-info) #\Tab)
            (ft "~A~%" (delim everything!))))))
    (info "finished conversion in ~A~%~%" (time-for-humans time!)))
  (gc :full t))


